Implement git tag checking for autoupdate.

jlaunonen opened the pull request:
For release:
- The **/package.json version field must be updated first to contain
correct version number and channel suffix (alpha, beta, release).
- A tag must be created with same values for the release commit,
e.g.  git tag -a -m "" 1.2.3-alpha
- Various packaging commands can be issued. Resulting
release/$channel-$os.yml file and executable must be copied to update
server.

For snapshot builds:
- The **/package.json version field must be updated to contain
"snapshot" as the channel suffix.
- Various packaging commands can be issued.

The package.json version is compared against tag name so they must
match.

jlaunonen committed:
Implement git tag checking for autoupdate.

For release:
- The **/package.json version field must be updated first to contain
correct version number and channel suffix (alpha, beta, release).
- A tag must be created with same values for the release commit,
e.g.  git tag -a -m "" 1.2.3-alpha
- Various packaging commands can be issued. Resulting
release/$channel-$os.yml file and executable must be copied to update
server.

For snapshot builds:
- The **/package.json version field must be updated to contain
"snapshot" as the channel suffix.
- Various packaging commands can be issued.

The package.json version is compared against tag name so they must
match.

mrinnetmaki commented on electron-builder-publish.js:
```diff
@@ -93,33 +95,71 @@ const config = {
-      provider: 's3',
-      bucket: 'downloads.tidepool.org',
+      provider: 'generic',
+      url: 'https://www.sensotrend.fi/download/uploader/update/${os}/',
```
Would there be a way to make this dependent on the channel?
dev.sensotrend.fi for dev, test.sensotrend.fi for test, www.sensotrend.fi for production?

jlaunonen replied:
Now possible via 7a3c827.

mrinnetmaki reviewed (approved):
> A tag must be created with same values for the release commit,
> e.g. git tag -a -m "" 1.2.3-alpha

Any way to automate this? That is, to get the version number and channel from the package.json, for instance?
Just thinking of avoiding possible typos...

jlaunonen committed:
Implement package.json autoupdate.

Gather the version, name, description and other information from git
describe.

jlaunonen commented:
> > A tag must be created with same values for the release commit,
> > e.g. git tag -a -m "" 1.2.3-alpha
> 
> Any way to automate this? That is, to get the version number and channel from the package.json, for instance?
> Just thinking of avoiding possible typos...

Yes, now in a279e3c. Just the reverse, though, as version information in package.json is outdated in conflicting manner due parallel development that keeping it there is not a good option. So, the version is not determined automatically depending on where the previous tag exists. If current commit doesn't have a tag, it must be a development run or snapshot build. And if current commit is tagged, it must be a build aimed for release that is declared by the tag. (If running development build in a tagged commit, it should not matter that the run/build is marked by the tag. Only thing is, that the version is not currently marked dirty if there are local uncommitted changes on a tagged commit.)

jlaunonen committed:
Refactor default environment selection and URL overrides.

By this, each server address is defined only once and does not require
manual tuning depending on whether the build is meant for development,
testing or production. Also the default server group can be changed with
a single variable rather than requiring to supply them all.
Additionally, the change does not pollute the pre-defined server groups.

Build-time environment variables:

- ENVIRONMENT configures which environment updates are checked from and
which server URL group is used by default.
- DEFAULT_ENVIRONMENT changes the default server URL group from what is
configured by definition set by ENVIRONMENT (or default).

Run-time environment variables:

- UPLOADER_ENVIRONMENT changes the default server URL group from what
would be used by default (defined in build-time by above variables).
- API_URL et co change the value from what would be used by the server
URL group. If any of these are set, a Custom entry will be added to
server selection menu.

jlaunonen committed:
Merge remote-tracking branch 'origin/sensotrend-uploader' into autoupdate

jlaunonen committed:
Extract server configurations to separate file.

mrinnetmaki reviewed (approved):
Good stuff!
I'm a bit concerned over how we'll try to get these changes adopted upstream (in Tidepool's repo). Probably one change at a time (environment management and tagged versions separately)? But we can worry about that later on.
@jlaunonen will you present this in practice in a demo session?
